FABulous/fabric_cad/timing_model/hdlnx/sdfnx/sdf_to_graph.py

```diff
@@ -0,0 +1,201 @@
+        current_pin: str = hier_pin_path
+        for _ in range(num_follow):
+            successors = next(self.graph.successors(current_pin), None)
+            if successors is None:
+                break
+            current_pin = successors
+        return current_pin
+    
+
+    def path_to_nearest_target_sentinel(
+        self,
+        source: str,
+        targets: list[str],
+        weight: str | None = None,
+        sentinel_prefix: str = "_sentinel_",
+        reverse: bool = False
+    ) -> tuple[list[str], str, float]:
+        """
+        Find the shortest path from `source` to the nearest node in `targets`
+        in a (directed) NetworkX graph using the sentinel-node trick.
+        https://networkx.org/documentation/stable/reference/algorithms/shortest_paths.html
+
+        Parameters
+        ----------
+        source : str
+            Source node.
+        targets : list[str]
+            List of target nodes.
+        weight : str | None, optional
+            Edge attribute name to use as weight. If None, the graph is treated
+            as unweighted (hop count).
+        sentinel_prefix : str, optional
+            Base name for the temporary sentinel node (ensured to be unique).
+
+        Returns
+        -------
+        path : list[str] | None
+            List of nodes from `source` to the closest target (no sentinel),
+            or None if no target is reachable.
+        closest_target : str | None
+            The closest target node, or None if no target is reachable.
+        distance : int | float | None
+            Distance from `source` to `closest_target`. For unweighted graphs
+            this is hop count; for weighted graphs, sum of edge weights.
+            None if no target is reachable.
+            
+        Raises
+        ------
+        ValueError
+            If `targets` is empty.
+        """
+        
+        if reverse:
+            G = self.graph.reverse(copy=False)
+        else:
+            G = self.graph
+        
+        targets: set[str] = set(targets)
+        if not targets:
+            raise ValueError("targets must be a non-empty iterable of nodes")
+
+        # Pick a sentinel name that doesn't collide with existing nodes
+        sentinel: str = f"{sentinel_prefix}_i89f9j9g58f7g6e5d4c3b2a1"
+
+        G.add_node(sentinel)
+
+        # Add zero-cost edges from each target to the sentinel
+        if weight is None:
+            for t in targets:
+                #if G.has_node(t):
+                G.add_edge(t, sentinel)
+        else:
+            zero_attr = {weight: 0}
+            for t in targets:
+                #if G.has_node(t):
+                G.add_edge(t, sentinel, **zero_attr)
+
+        try:
+            # Shortest path (directed) source -> sentinel
+            path: list[str] = nx.shortest_path(G, source=source, target=sentinel, weight=weight)
+            dist: float = nx.shortest_path_length(G, source=source, target=sentinel, weight=weight)
```

Comment:
Since you say the runtime is 20 mins, you can do [pairwise](https://docs.python.org/3/library/itertools.html#itertools.pairwise) over the path and get the edge from the graph, and sum out the weight. This helper should help [get_edge_data](https://networkx.org/documentation/stable/reference/classes/generated/networkx.DiGraph.get_edge_data.html#networkx.DiGraph.get_edge_data). Since you will not be calculating the path twice.